craft_parts/executor/part_handler.py

```diff
@@ -475,6 +482,21 @@ def _update_build(self, step_info: StepInfo) -> None:
 
         self._organize(overwrite=True)
 
+    def _reapply_action(self, action: Action, *, step_info: StepInfo) -> None:
+        """Call the appropriate reapply handler for the given step."""
+        if action.step == Step.OVERLAY:
+            self._reapply_overlay(step_info)
+        else:
+            step_name = action.step.name.lower()
+            raise errors.InvalidAction(
+                f"cannot reapply step {step_name!r} of {self._part.name!r}"
+            )
+
+    def _reapply_overlay(self, step_info) -> None:
+        """Clean and repopulate the current part's layer, keeping its state."""
+        shutil.rmtree(self._part.part_layer_dir)
+        self._run_overlay(step_info)
+
```

Comment:
This is the actual layer reapplication on disk: all previous contents are removed and the layer is completely re-populated.